### PR TITLE
TreeDataProvider (json outline for deployment templates, param files and createUiDefinition

### PR DESCRIPTION
**Pull request by marcvaneijk**
- Updated parser to JSON.ts
- Conditional icons on resources
- expand nested elements
- Key Value pair content as label on string elements in outline
- Select to navigate code
- Fixed undefined when switching ActiveTextEditor
- Only show outline explorer when active language mode is json, close if other language
- Fixed undefined when using promise for treeItem results
- Added both the key and the value in selection navigation
- Elements for index are presented as json object string.
- Override for object in array labels. When the name element is present, it is used as the label for the object. 

**Please note:** 
- Verify if the location of the registerTreeDataProvider needs to be updated:  https://github.com/Microsoft/vscode-azurearmtools/pull/15/files#diff-a7c865ae04d74e17a62b5922a2a19973R77
- Not using type on tree object : https://github.com/Microsoft/vscode-azurearmtools/pull/15/files#diff-0b7fcb613ac6712f4c9b9486e871e667R7 Setting the type to ParseResult from JSON.ts will result in the value.properties element as private. I need the value.properties to resolve the indexnumbers. I didn't want to tamper with the existing interface in JSON.ts, so I have not used a type on the tree object.

### CODE REVIEW

**Comment by whiskeyjay** [comment]
Great addition to the extension! 👍

I noticed the TreeView works for any JSON files, just if it is not an ARM template, it doesn't look very well because lack of icons. Do you think we should limit the TreeView for ARM templates only?

File: src/Treeview.ts [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+                    else if (elementInfo.current.value.type === "ArrayValue" && elementInfo.current.collapsible) {
+                        // Array with Object
+                        if (valueNode.elements[0].constructor.name === "ObjectValue"){
+                            for(var i=0, il=valueNode.length; i < il; i++){
+                                let item = this.getElementInfo(valueNode.elements[i], elementInfo);
+                                result.push(item);
+                            }
+                        }
+                        // Array contains other values
+                        else {}
```
[the remark this answers is not in the document]
Comment (marcvaneijk) [reply src/Treeview.ts]: removed empty else

File: src/Treeview.ts [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+            // Center the method in the document
+            editor.revealRange(range, vscode.TextEditorRevealType.Default);
+            // Select the method name
+            editor.selection = new vscode.Selection(range.start, range.end);
+            // Swap the focus to the editor
+            vscode.window.showTextDocument(editor.document, editor.viewColumn, false);
+    }
+
+    private parseTree(document?: vscode.TextDocument): void {
+        document = document !== undefined ? document : vscode.window.activeTextEditor.document;
```
[the remark this answers is not in the document]
Comment (marcvaneijk) [reply src/Treeview.ts]: Changed condition to only parseTree if the languageID === "json"

File: src/AzureRMTools.ts [hunk 3]
```diff
@@ -72,6 +73,10 @@ export class AzureRMTools {
     constructor() {
         this.loadConfiguration();
 
+        const jsonOutline = new JsonOutlineProvider();
+        vscode.window.registerTreeDataProvider("json-outline", jsonOutline);
```
[the remark this answers is not in the document]
Comment (marcvaneijk) [reply src/AzureRMTools.ts]: I can use a little help on this one. Not sure what you would like me to do here..

**Comment by marcvaneijk** [comment]
Thanks @whiskeyjay, I've coded the treeview explicitly for ARM deployment templates, param files and createUiDefinition. This gives it a better experience for this specific extension. Actually there is a more generic extension for different languages (also json) available here https://marketplace.visualstudio.com/items?itemName=DaGhostman.vs-treeview

File: src/AzureRMTools.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (whiskeyjay) [reply src/AzureRMTools.ts]: No worries. I'll take care of it later.

**Pull request merged**